From 1d8ca76508dbc98f5ecd348f14714cb8b9072d09 Mon Sep 17 00:00:00 2001
From: Shadi Ammouri <shadi@marvell.com>
Date: Sun, 6 Mar 2016 11:27:17 +0200
Subject: [PATCH 041/239] palladium: Enable MPP configuration under palladium

This is needed in order to output printouts to the terminal.
In normal systems, the MPP configuration is done by Bootrom.

Change-Id: I7ca147b3a12addba5e7547631f691d07ae411d5e
Signed-off-by: Shadi Ammouri <shadi@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/28048
Reviewed-by: Haim Boot <hayim@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
---
 include/plat/marvell/a8k/common/plat_marvell.h |  3 ++
 plat/marvell/a8k/a7040_rz/plat_bl1_setup.c     | 49 ++++++++++++++++++++++++++
 plat/marvell/a8k/a7040_rz/platform.mk          |  4 +++
 plat/marvell/common/marvell_bl1_setup.c        |  5 +++
 4 files changed, 61 insertions(+)
 create mode 100644 plat/marvell/a8k/a7040_rz/plat_bl1_setup.c

diff --git a/include/plat/marvell/a8k/common/plat_marvell.h b/include/plat/marvell/a8k/common/plat_marvell.h
index b343c37..c050a7c 100644
--- a/include/plat/marvell/a8k/common/plat_marvell.h
+++ b/include/plat/marvell/a8k/common/plat_marvell.h
@@ -127,5 +127,8 @@ int plat_marvell_get_alt_image_source(
 	uintptr_t *image_spec);
 unsigned int plat_marvell_calc_core_pos(u_register_t mpidr);
 
+#if PALLADIUM
+void marvell_bl1_setup_mpps(void);
+#endif
 
 #endif /* __PLAT_MARVELL_H__ */
diff --git a/plat/marvell/a8k/a7040_rz/plat_bl1_setup.c b/plat/marvell/a8k/a7040_rz/plat_bl1_setup.c
new file mode 100644
index 0000000..1cdc2bc
--- /dev/null
+++ b/plat/marvell/a8k/a7040_rz/plat_bl1_setup.c
@@ -0,0 +1,49 @@
+/*
+* ***************************************************************************
+* Copyright (C) 2016 Marvell International Ltd.
+* ***************************************************************************
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+*
+* Redistributions of source code must retain the above copyright notice, this
+* list of conditions and the following disclaimer.
+*
+* Redistributions in binary form must reproduce the above copyright notice,
+* this list of conditions and the following disclaimer in the documentation
+* and/or other materials provided with the distribution.
+*
+* Neither the name of Marvell nor the names of its contributors may be used
+* to endorse or promote products derived from this software without specific
+* prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+* ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+* LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+* OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+* POSSIBILITY OF SUCH DAMAGE.
+*
+***************************************************************************
+*/
+
+#include <plat_marvell.h>
+#include <mmio.h>
+#include <apn806_setup.h>
+#include <cp110_setup.h>
+
+#define MVEBU_MPP_REGS(n)	(0x6F4000 + ((n) << 2))
+void marvell_bl1_setup_mpps(void)
+{
+	/* Enable UART MPPs.
+	** In a normal system, this is done by Bootrom.
+	*/
+	mmio_write_32(MVEBU_REGS_BASE + MVEBU_MPP_REGS(1), 0x3000);
+	mmio_write_32(MVEBU_REGS_BASE + MVEBU_MPP_REGS(2), 0x3000);
+}
+
diff --git a/plat/marvell/a8k/a7040_rz/platform.mk b/plat/marvell/a8k/a7040_rz/platform.mk
index 669b49a..c1c2880 100644
--- a/plat/marvell/a8k/a7040_rz/platform.mk
+++ b/plat/marvell/a8k/a7040_rz/platform.mk
@@ -68,6 +68,10 @@ PLAT_BL_COMMON_SOURCES	:=	$(PLAT_SRC_BASE)/aarch64/a7040_rz_common.c	\
 BL1_SOURCES		+=	$(PLAT_SRC_BASE)/aarch64/plat_helpers.S	\
 				lib/cpus/aarch64/cortex_a72.S
 
+ifeq (${PALLADIUM}, 1)
+BL1_SOURCES		+=	$(PLAT_SRC_BASE)/plat_bl1_setup.c
+endif
+
 MARVELL_DRV		:= 	$(MARVELL_DRV_BASE)/rfu.c	\
 				$(MARVELL_DRV_BASE)/iob.c	\
 				$(MARVELL_DRV_BASE)/ccu.c
diff --git a/plat/marvell/common/marvell_bl1_setup.c b/plat/marvell/common/marvell_bl1_setup.c
index 8b5e10f..e68667c 100644
--- a/plat/marvell/common/marvell_bl1_setup.c
+++ b/plat/marvell/common/marvell_bl1_setup.c
@@ -79,6 +79,11 @@ void marvell_bl1_early_platform_setup(void)
 {
 	const size_t bl1_size = BL1_RAM_LIMIT - BL1_RAM_BASE;
 
+#if PALLADIUM
+	/* Initialize SoC MPPs in case it's needed. */
+	marvell_bl1_setup_mpps();
+#endif
+
 	/* Initialize the console to provide early debug support */
 	console_init(PLAT_MARVELL_BOOT_UART_BASE, PLAT_MARVELL_BOOT_UART_CLK_IN_HZ,
 			MARVELL_CONSOLE_BAUDRATE);
-- 
1.9.1

